### PR TITLE
[SPARK-13306][SQL] Addendum to uncorrelated scalar subquery

### PR DESCRIPTION
**Pull request by rxin**
## What changes were proposed in this pull request?

This pull request fixes some minor issues (documentation, test flakiness, test organization) with #11190, which was merged earlier tonight.
## How was the this patch tested?

unit tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala [hunk 1]
```diff
@@ -272,14 +272,3 @@ case class Literal protected (value: Any, dataType: DataType)
     case _ => value.toString
   }
 }
-
-// TODO: Specialize
-case class MutableLiteral(var value: Any, dataType: DataType, nullable: Boolean = true)
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala]: this was unused. just deleting it.

**Comment by rxin** [comment]
cc @davies 
can you explain what the following test case is testing?

```
     assertResult(Array(Row(1))) {
       sql("with t2 as (select 1 as b, 2 as c) " +
         "select a from (select 1 as a union all select 2 as a) t " +
         "where a = (select max(b) from t2) ").collect()
     }
```

**Comment by rxin** [comment]
Also explain why wholestagecodegen needs to call waitforsubqueries explicitly.

File: sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala [hunk 2]
```diff
@@ -36,36 +36,39 @@ class SubquerySuite extends QueryTest with SharedSQLContext {
...
     assertResult(Array(Row(null))) {
       sql("select (select 's' as s limit 0) as b").collect()
     }
   }
 
+  test("analysis error when the number of columns is not 1") {
+    val error = intercept[AnalysisException] {
+      sql("select (select 1, 2) as b").collect()
+    }
+    assert(error.message.contains("Scalar subquery must return only one column, but got 2"))
```
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala]: These are already covered by AnasisErrorSuite, some one said that we don't need to cover every corner cases by end-to-end query tests.

**Comment by davies** [comment]
For those operators that is generated in WholeStageCodegen, the execute() of them are not called, but produce() will be called, we should make sure that all the subqueries all executed before generating the code in doProduce() and doConsume()

**Comment by davies** [comment]
That case is used to test CTE and aggregation inside subquery.

**Comment by rxin** [comment]
Merging this in master.

**Pull request closed without merging**